pageserver/src/tenant/blob_io.rs

```diff
@@ -83,41 +84,109 @@ impl<'a> BlockCursor<'a> {
     }
 }
 
+/// A wrapper of `VirtualFile` that allows users to write blobs.
 ///
-/// An implementation of BlobWriter to write blobs to anything that
-/// implements std::io::Write.
-///
-pub struct WriteBlobWriter<W> {
-    inner: W,
+/// If a `BlobWriter` is dropped, the internal buffer will be
+/// discarded. You need to call [`flush_buffer`](Self::flush_buffer)
+/// manually before dropping.
+pub struct BlobWriter<const BUFFERED: bool> {
+    inner: VirtualFile,
     offset: u64,
+    /// A buffer to save on read calls
+    buf: [u8; PAGE_SZ],
+    /// The number of bytes already occupied in buf
+    /// In other words: pointer to the first unwritten byte in buf.
+    ///
+    /// After each `write_all` call concludes, we maintain the
+    /// invariant that buf_offs < buf.len(), so the buffer is
+    /// never completely full outside of the `write_all` function.
+    buf_offs: usize,
 }
 
-impl<W> WriteBlobWriter<W> {
-    pub fn new(inner: W, start_offset: u64) -> Self {
-        WriteBlobWriter {
+impl<const BUFFERED: bool> BlobWriter<BUFFERED> {
+    pub fn new(inner: VirtualFile, start_offset: u64) -> Self {
+        Self {
             inner,
             offset: start_offset,
+            buf: [0; PAGE_SZ],
```

Comment:
Have you considered using Cursor+Vec here, instead of manual buffer+offset?  I think the efficiency would be similar, and it would need a bit less careful testing on our end.

Reply:
I've adopted a `Vec` now.